File: packages/web/src/lib/store.ts
import { create } from 'zustand'
import { api, Route, AuthRule } from './api'

interface Store {
  routes: Route[]
  authRules: AuthRule[]
  loading: boolean
  error: string | null
  fetchRoutes: () => Promise<void>
  fetchAuthRules: () => Promise<void>
  createRoute: (data: Partial<Route>) => Promise<void>
  updateRoute: (id: string, data: Partial<Route>) => Promise<void>
  deleteRoute: (id: string) => Promise<void>
  createAuthRule: (data: Partial<AuthRule>) => Promise<void>
  updateAuthRule: (id: string, data: Partial<AuthRule>) => Promise<void>
  deleteAuthRule: (id: string) => Promise<void>
}

export const useStore = create<Store>((set, get) => ({
  routes: [],
  authRules: [],
  loading: false,
  error: null,

  fetchRoutes: async () => {
    set({ loading: true, error: null })
    try {
      const routes = await api.listRoutes()
      set({ routes, loading: false })
    } catch (e) {
      set({ error: (e as Error).message, loading: false })
    }
  },

  fetchAuthRules: async () => {
    try {
      const authRules = await api.listAuthRules()
      set({ authRules })
    } catch (e) {
      set({ error: (e as Error).message })
    }
  },

  createRoute: async (data) => {
    const route = await api.createRoute(data)
    set({ routes: [...get().routes, route] })
  },

  updateRoute: async (id, data) => {
    const route = await api.updateRoute(id, data)
    set({ routes: get().routes.map(r => r.id === id ? route : r) })
  },

  deleteRoute: async (id) => {
    await api.deleteRoute(id)
    set({ routes: get().routes.filter(r => r.id !== id) })
  },

  createAuthRule: async (data) => {
    const rule = await api.createAuthRule(data)
    set({ authRules: [...get().authRules, rule] })
  },

  updateAuthRule: async (id, data) => {
    const rule = await api.updateAuthRule(id, data)
    set({ authRules: get().authRules.map(r => r.id === id ? rule : r) })
  },

  deleteAuthRule: async (id) => {
    await api.deleteAuthRule(id)
    set({ authRules: get().authRules.filter(r => r.id !== id) })
  },
}))

File: packages/server/internal/store/store.go
package store

import "time"

// Store is the persistence interface for all domain entities.
type Store interface {
	// Routes
	ListRoutes() ([]Route, error)
	GetRoute(id string) (*Route, error)
	CreateRoute(r *Route) error
	UpdateRoute(r *Route) error
	DeleteRoute(id string) error

	// Auth rules (legacy, kept for migration)
	ListAuthRules() ([]AuthRule, error)
	GetAuthRule(id string) (*AuthRule, error)
	GetAuthRuleByRouteID(routeID string) (*AuthRule, error)
	CreateAuthRule(r *AuthRule) error
	UpdateAuthRule(r *AuthRule) error
	DeleteAuthRule(id string) error

	// Route Auth Config
	GetRouteAuthConfig(routeID string) (*RouteAuthConfig, error)
	CreateOrUpdateRouteAuthConfig(c *RouteAuthConfig) error
	DeleteRouteAuthConfig(routeID string) error

	// API Keys
	ListApiKeysByRoute(routeID string) ([]ApiKey, error)
	GetApiKey(id string) (*ApiKey, error)
	FindApiKeyBySecret(secret string) (*ApiKey, error)
	CreateApiKey(k *ApiKey) error
	UpdateApiKey(k *ApiKey) error
	DeleteApiKey(id string) error

	// Users
	ListUsers() ([]User, error)
	GetUserByUsername(username string) (*User, error)
	GetUserByID(id string) (*User, error)
	CreateUser(u *User) error
	UpdateUser(u *User) error
	DeleteUser(id string) error
	VerifyPassword(user *User, password string) bool
	EnsureAdmin(username, password string) (bool, error)
	HasAdminUsers() (bool, error)

	// Certificates
	ListCertificates() ([]Certificate, error)
	GetCertificate(id string) (*Certificate, error)
	GetCertificateByDomain(domain string) (*Certificate, error)
	CreateCertificate(c *Certificate) error
	UpdateCertificate(c *Certificate) error
	DeleteCertificate(id string) error
	ListExpiringLocalCertificates(before time.Time) ([]Certificate, error)

	// CA certificates
	ListCACertificates() ([]CACertificate, error)
	GetCACertificate(id string) (*CACertificate, error)
	GetFirstCACertificate() (*CACertificate, error)
	CreateCACertificate(c *CACertificate) error

	// Host profiles
	ListHostProfiles() ([]HostProfile, error)
	GetHostProfile(id string) (*HostProfile, error)
	CreateHostProfile(p *HostProfile) error
	UpdateHostProfile(p *HostProfile) error
	DeleteHostProfile(id string) error
	SetActiveHostProfile(profileID string) error

	// Host entries
	CreateHostEntry(e *HostEntry) error
	ListHostEntries(profileID string) ([]HostEntry, error)
	GetHostEntry(id string) (*HostEntry, error)
	UpdateHostEntry(e *HostEntry) error
	DeleteHostEntry(id string) error
	ReorderHostEntries(profileID string, orderedIDs []string) error
	ListEnabledHostEntries(profileID string) ([]HostEntry, error)

	// Lifecycle
	Close() error
}
